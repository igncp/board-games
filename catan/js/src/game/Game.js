// @flow

import {createPlayer} from "../Player"
import type {T_Player} from "../Player"
import {
  createBoard,
  assignCitiesToPlayer,
  assignSettlementsToPlayer,
  assignRoadsToPlayer,
  placeCityForPlayer as placeCityForPlayerInBoard,
  placeSettlementForPlayer as placeSettlementForPlayerInBoard,
  placeRoadForPlayer as placeRoadForPlayerInBoard,
} from "../Board"
import type {T_CityPosition, T_SettlementPosition, T_RoadPosition} from "../Board"

import type {T_Game} from "./GameCore"
import {createTimeline} from "./Timeline"

type T_createGame = ({|
  playersNumber: number
|}) => T_Game

const createPlayers = (playersNumber) => {
  const players = []

  for (let i = 0; i < playersNumber; i++) {
    const newPlayer = createPlayer({
      id: i.toString(),
      excludedColors: players.map((p) => p.color),
    })

    players.push(newPlayer)
  }

  return players
}

const CITIES_PER_PLAYER = 4
const SETTLEMENTS_PER_PLAYER = 5
const ROADS_PER_PLAYER = 15

const assignPiecesToPlayers = (players, board) => {
  players.forEach((player) => {
    const commonOpts = {
      playerReference: player.id,
      board,
    }

    assignCitiesToPlayer({
      ...commonOpts,
      citiesNumber: CITIES_PER_PLAYER,
    })

    assignSettlementsToPlayer({
      ...commonOpts,
      settlementsNumber: SETTLEMENTS_PER_PLAYER,
    })

    assignRoadsToPlayer({
      ...commonOpts,
      roadsNumber: ROADS_PER_PLAYER,
    })
  })
}

export const createGame: T_createGame = ({playersNumber}) => {
  const timeline = createTimeline()
  const players = createPlayers(playersNumber)
  const board = createBoard({
    playersNumber: players.length,
    roadsPerPlayer: ROADS_PER_PLAYER,
    settlementsPerPlayer: SETTLEMENTS_PER_PLAYER,
    citiesPerPlayer: CITIES_PER_PLAYER,
  })

  assignPiecesToPlayers(players, board)

  return {
    board,
    players,
    timeline,
  }
}

type T_placeCityForPlayer = ({|
  cityPosition: T_CityPosition,
  game: T_Game,
  player: T_Player,
|}) => void

export const placeCityForPlayer: T_placeCityForPlayer = ({game, player, cityPosition}) => {
  placeCityForPlayerInBoard({
    board: game.board,
    playerReference: player.id,
    cityPosition,
  })
}

type T_placeSettlementForPlayer = ({|
  game: T_Game,
  player: T_Player,
  settlementPosition: T_SettlementPosition,
|}) => void

export const placeSettlementForPlayer: T_placeSettlementForPlayer = ({game, player, settlementPosition}) => {
  placeSettlementForPlayerInBoard({
    board: game.board,
    playerReference: player.id,
    settlementPosition,
  })
}

type T_placeRoadForPlayer = ({|
  game: T_Game,
  player: T_Player,
  roadPosition: T_RoadPosition,
|}) => void

export const placeRoadForPlayer: T_placeRoadForPlayer = ({game, player, roadPosition}) => {
  placeRoadForPlayerInBoard({
    board: game.board,
    playerReference: player.id,
    roadPosition,
  })
}
