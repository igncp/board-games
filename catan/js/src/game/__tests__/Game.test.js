// @flow

import * as utilsMod from "../../Utils"

import {createGame, placeCityForPlayer, placeRoadForPlayer} from "../Game"

jest.spyOn(utilsMod, "getRandomItem")

describe("Game", () => {
  beforeEach(() => {
    (utilsMod.getRandomItem: any).mockImplementation((arr) => arr[0])
  })

  describe("createGame", () => {
    it("returns the expected data", () => {
      (utilsMod.getRandomItem: any).mockImplementation((arr) => arr[0])

      const result = createGame({
        playersNumber: 4,
      })

      expect(result).toMatchSnapshot()
    })
  })

  describe("placeCityForPlayer", () => {
    it("updates the board with expected values", () => {
      const game = createGame({
        playersNumber: 4,
      })

      placeCityForPlayer({
        game,
        cityPosition: ["06", "01", "02"],
        player: game.players[0],
      })

      expect(game.board.positionToCityMap).toMatchSnapshot()
    })
  })

  describe("placeRoadForPlayer", () => {
    it("errors if position is in water", () => {
      const game = createGame({
        playersNumber: 4,
      })

      const fn = () => placeRoadForPlayer({
        game,
        roadPosition: ["00", "01"],
        player: game.players[0],
      })

      expect(fn).toThrow("The road can not be placed in water")
    })
  })
})
