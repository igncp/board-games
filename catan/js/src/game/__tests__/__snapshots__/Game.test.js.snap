// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game createGame returns the expected data 1`] = `
Object {
  "board": Object {
    "citiesToPlayerMap": Object {
      "0": "0",
      "1": "0",
      "10": "2",
      "11": "2",
      "12": "3",
      "13": "3",
      "14": "3",
      "15": "3",
      "2": "0",
      "3": "0",
      "4": "1",
      "5": "1",
      "6": "1",
      "7": "1",
      "8": "2",
      "9": "2",
    },
    "grid": Object {
      "chits": Object {
        "05": "robber",
        "06": "10",
        "07": "10",
        "10": "11",
        "11": "11",
        "12": "12",
        "13": "02",
        "16": "03",
        "17": "03",
        "18": "04",
        "19": "04",
        "20": "05",
        "23": "05",
        "24": "06",
        "25": "06",
        "26": "08",
        "29": "08",
        "30": "09",
        "31": "09",
      },
      "edges": Object {},
      "sides": Object {},
      "terrains": Object {
        "00": "water",
        "01": "water",
        "02": "water",
        "03": "water",
        "04": "water",
        "05": "hill",
        "06": "hill",
        "07": "hill",
        "08": "water",
        "09": "water",
        "10": "pasture",
        "11": "pasture",
        "12": "pasture",
        "13": "pasture",
        "14": "water",
        "15": "water",
        "16": "mountain",
        "17": "mountain",
        "18": "mountain",
        "19": "field",
        "20": "field",
        "21": "water",
        "22": "water",
        "23": "field",
        "24": "field",
        "25": "forest",
        "26": "forest",
        "27": "water",
        "28": "water",
        "29": "forest",
        "30": "forest",
        "31": "desert",
        "32": "water",
        "33": "water",
        "34": "water",
        "35": "water",
        "36": "water",
      },
    },
    "positionToCityMap": Object {},
    "positionToRoadMap": Object {},
    "positionToSettlementMap": Object {},
    "roadToPlayerMap": Object {
      "0": "0",
      "1": "0",
      "10": "0",
      "11": "0",
      "12": "0",
      "13": "0",
      "14": "0",
      "15": "1",
      "16": "1",
      "17": "1",
      "18": "1",
      "19": "1",
      "2": "0",
      "20": "1",
      "21": "1",
      "22": "1",
      "23": "1",
      "24": "1",
      "25": "1",
      "26": "1",
      "27": "1",
      "28": "1",
      "29": "1",
      "3": "0",
      "30": "2",
      "31": "2",
      "32": "2",
      "33": "2",
      "34": "2",
      "35": "2",
      "36": "2",
      "37": "2",
      "38": "2",
      "39": "2",
      "4": "0",
      "40": "2",
      "41": "2",
      "42": "2",
      "43": "2",
      "44": "2",
      "45": "3",
      "46": "3",
      "47": "3",
      "48": "3",
      "49": "3",
      "5": "0",
      "50": "3",
      "51": "3",
      "52": "3",
      "53": "3",
      "54": "3",
      "55": "3",
      "56": "3",
      "57": "3",
      "58": "3",
      "59": "3",
      "6": "0",
      "7": "0",
      "8": "0",
      "9": "0",
    },
    "settlementToPlayerMap": Object {
      "0": "0",
      "1": "0",
      "10": "2",
      "11": "2",
      "12": "2",
      "13": "2",
      "14": "2",
      "15": "3",
      "16": "3",
      "17": "3",
      "18": "3",
      "19": "3",
      "2": "0",
      "3": "0",
      "4": "0",
      "5": "1",
      "6": "1",
      "7": "1",
      "8": "1",
      "9": "1",
    },
  },
  "players": Array [
    Object {
      "color": "RED",
      "id": "0",
    },
    Object {
      "color": "BLUE",
      "id": "1",
    },
    Object {
      "color": "ORANGE",
      "id": "2",
    },
    Object {
      "color": "WHITE",
      "id": "3",
    },
  ],
  "timeline": Object {
    "phase": "INITIAL_SETUP",
  },
}
`;

exports[`Game placeCityForPlayer updates the board with expected values 1`] = `
Object {
  "01_02_06": "0",
}
`;
